juju/juju-controller
# Send signal to controller when controller config changes

This PR builds on #60 by swapping the SIGHUP for a socket call. The description of #60 is:

> This patch does two things:
> 
>  - We no longer assume that the identity of the controller itself is congruent with the controller unit ID. We now get this information from the running controller jujud process.
>  - When the controller configuration file is changed, we send a SIGHUP to the controller's process, which will be detected there, causing a reload of the new configuration.
> 
> To verify:
> ...   
> 
> https://warthogs.atlassian.net/browse/JUJU-5306

In this PR, it is changed so that instead of triggering a config reload on the controller by sending a `SIGHUP`, a `/reload` request is sent to the `controllerconfig` socket. This change is made because the `SIGHUP` signal is not reliable on Kubernetes.

The new QA is:
- `charmcraft pack` this charm.
- Build juju in branch https://github.com/juju/juju/pull/16961
- Bootstrap a new LXD controller using `--controller-charm-path <path-to-packed-artefact>`.
- `juju enable-ha`.
- Once all units have joined the relation, ssh to the controller machine and check _/var/lib/juju/agents/controller-0/agent.conf_. It should be populated along the lines of:
```
db-bind-addresses:
  controller/0: 10.246.27.181
  controller/1: 10.246.27.53
  controller/2: 10.246.27.12
```
- `juju debug-log -m controller --replay|grep "controlleragentconfig reload config request received, reloading config"` indicate the config has reloaded for each controller
- Deploy dashboard to check everything is working as expected `juju deploy juju-dashboard --channel beta` followed by `juju integrate controller juju-dashboard`
